File: src/dcs_mission_creator/core/lua/emcon.lua
-- dcs_mission_creator: SAM EMCON reaction to anti-radiation missiles.
--
-- Rendered by `core/emcon.py` into a mission-start DoScript. It fills three
-- placeholders (named here without their underscores so this comment is not
-- itself substituted): SITES is one `{name=…, prob=…, …}` row per site, SIDE
-- the coalition constant the radio calls go to, SPACING the seconds left
-- between two consecutive radio calls.
do
  local sites = {
__SITES__
  }
  local side = __SIDE__
  local spacing = __SPACING__
  local armNames = {"AGM_88", "AGM_45", "AGM_122", "ALARM", "Kh-25MP", "Kh-31P", "Kh-58"}
  local state = {}

  -- One HARM shot puts several sites dark at once, so calls are *queued* and
  -- played `spacing` seconds apart rather than dropped: every site still gets
  -- its own call, they just do not talk over each other. The queue is capped
  -- so a busy net cannot back up a minute of stale traffic.
  local queue, draining = {}, false

  local function playNext(_, time)
    local item = table.remove(queue, 1)
    if item == nil then
      draining = false
      return nil
    end
    if item.text then trigger.action.outTextForCoalition(side, item.text, 12) end
    if item.sound then trigger.action.outSoundForCoalition(side, item.sound) end
    return time + spacing
  end

  local function announce(text, sound, time)
    if text == nil and sound == nil then return end
    -- Sites that share wording (the EWR pair) say it once, not once each.
    for _, q in ipairs(queue) do
      if q.text == text and q.sound == sound then return end
    end
    if #queue >= 8 then table.remove(queue, 1) end
    queue[#queue + 1] = {text = text, sound = sound}
    if not draining then
      draining = true
      timer.scheduleFunction(playNext, {}, time + 0.1)
    end
  end

  -- Reaction time is not a flat coin flip across the band: the middle is the
  -- common case, and both a snap reaction and a badly slow one are rare.
  -- Averaging two uniform draws gives that triangular shape in one line. The
  -- same spread reads right for how long a crew then stays off the air.
  local function spread(lo, hi)
    return lo + (math.random() + math.random()) * 0.5 * (hi - lo)
  end

  -- A site that has lost its radars is not dark, it is dead — it has nothing
  -- left to switch off and the crew making the call is gone. `Group.isExist()`
  -- stays true while a single launcher or the command post survives, so the one
  -- thing the calls have to be gated on is a live emitter. These are the
  -- attributes DCS tags radar vehicles with (search, track/fire-control, EWR).
  local radarAttrs = {"SAM SR", "SAM TR", "EWR"}

  local function liveRadars(name)
    local g = Group.getByName(name)
    if not g or not g:isExist() then return 0 end
    local n = 0
    for _, u in ipairs(g:getUnits() or {}) do
      if u:isExist() and u:getLife() > 0 then
        for _, attr in ipairs(radarAttrs) do
          if u:hasAttribute(attr) then
            n = n + 1
            break
          end
        end
      end
    end
    return n
  end

  -- `site.radars` remembers the most radars this site was ever seen with, so a
  -- group activated after mission start is counted the first time it is looked
  -- at rather than mistaken for a wreck. A site that never had a radar-tagged
  -- unit at all (a hand-built group of some type DCS does not tag) falls back to
  -- plain group existence, so it still reacts instead of going silent forever.
  local function radiating(site)
    local n = liveRadars(site.name)
    if n > (site.radars or 0) then site.radars = n end
    if n > 0 then return true end
    if (site.radars or 0) > 0 then return false end
    local g = Group.getByName(site.name)
    return g ~= nil and g:isExist()
  end

  -- Counted here, at mission start, while every site is still intact: a radar
  -- killed by something other than a HARM, before the first anti-radiation shot
  -- of the mission, has to read as dead rather than as one of those untagged
  -- groups. Late-activated sites do not exist yet and are picked up lazily above.
  for _, site in ipairs(sites) do
    site.radars = liveRadars(site.name)
  end

  local function emissions(site, on)
    local g = Group.getByName(site.name)
    if not g or not g:isExist() then return false end
    if not radiating(site) then return false end
    local c = g:getController()
    if on then
      c:setOption(AI.Option.Ground.id.ALARM_STATE, AI.Option.Ground.val.ALARM_STATE.RED)
      c:setOption(AI.Option.Ground.id.ROE, AI.Option.Ground.val.ROE.OPEN_FIRE)
    else
      c:setOption(AI.Option.Ground.id.ALARM_STATE, AI.Option.Ground.val.ALARM_STATE.GREEN)
      c:setOption(AI.Option.Ground.id.ROE, AI.Option.Ground.val.ROE.WEAPON_HOLD)
    end
    return true
  end

  local function wakeUp(site, time)
    local st = state[site.name]
    -- A later launch pushed the wake-up back; sleep on until then.
    if st and st.wakeAt > time + 0.5 then return st.wakeAt end
    state[site.name] = nil
    if emissions(site, true) then
      announce(site.upText, site.upSound, time)
    end
    return nil
  end

  local function goDark(site, time)
    local down = spread(site.downMin, site.downMax)
    local st = state[site.name]
    if st then
      -- Already dark: repeated fire keeps the crew off the air longer.
      st.wakeAt = math.max(st.wakeAt, time + down)
      return nil
    end
    if not emissions(site, false) then return nil end
    state[site.name] = {wakeAt = time + down}
    announce(site.downText, site.downSound, time)
    timer.scheduleFunction(wakeUp, site, time + down)
    return nil
  end

  local function isArm(w)
    if w == nil then return false end
    local ok, desc = pcall(function() return w:getDesc() end)
    if not ok or desc == nil then return false end
    if Weapon and Weapon.GuidanceType and desc.guidance == Weapon.GuidanceType.RADAR_PASSIVE then
      return true
    end
    local tn = desc.typeName or ""
    for _, pat in ipairs(armNames) do
      if string.find(tn, pat, 1, true) then return true end
    end
    return false
  end

  local handler = {}
  function handler:onEvent(event)
    if event == nil or event.id ~= world.event.S_EVENT_SHOT then return end
    if not isArm(event.weapon) then return end
    local shooter = event.initiator
    if shooter == nil or not shooter:isExist() then return end
    local sp = shooter:getPoint()
    local now = timer.getTime()
    for _, site in ipairs(sites) do
      local g = Group.getByName(site.name)
      if g and g:isExist() and radiating(site) then
        local units = g:getUnits()
        local u = units and units[1]
        if u then
          local p = u:getPoint()
          local dx, dz = sp.x - p.x, sp.z - p.z
          if math.sqrt(dx * dx + dz * dz) <= site.range and math.random() <= site.prob then
            -- Drawn per site per shot, and deliberately of the same order as a
            -- HARM's time of flight: the missile's range at launch is what
            -- decides whether it arrives before the transmitter dies.
            timer.scheduleFunction(goDark, site, now + spread(site.delayMin, site.delayMax))
          end
        end
      end
    end
  end
  world.addEventHandler(handler)
end

File: src/dcs_mission_creator/core/emcon.py
"""Radar EMCON reactions to anti-radiation missiles (project-owned helper).

Out of the box a DCS SAM site keeps its radar lit while a HARM rides the beam
all the way in: the crew never reacts, so every anti-radiation shot is a
guaranteed kill and SEAD degenerates into a shooting gallery. Real crews do the
opposite — the launch is called over the IADS net, the fire-control radar drops
emissions within seconds, the missile loses its emitter and goes for the last
known point, and the site comes back up several minutes later once the shooter
is assumed dry.

`arm_emcon_reaction` builds that behaviour as a mission-start `DoScript`: one
Lua event handler on `S_EVENT_SHOT` that recognises anti-radiation weapons,
decides per site whether that crew catches the launch, and cycles the site
through `ALARM_STATE GREEN` (radars off, weapons hold) and back to
`ALARM_STATE RED`. What makes it read as crew behaviour rather than a switch:

- **Not every crew reacts.** Each site has its own `probability`; a green crew
  keeps radiating and eats the missile.
- **Reaction takes time.** `delay_s` seconds pass between launch and emissions
  drop, so a HARM already close still kills.
- **The site comes back.** `shutdown_s` seconds later the radar re-radiates —
  suppression is temporary, destruction is not.
- **Repeat fire makes crews shy.** A second launch while a site is dark extends
  the shutdown instead of restarting it.
- **Range gates the reaction.** Only sites within `react_range_m` of the
  shooter hear about the launch; a HARM fired at one end of the map does not
  shut down the whole theater.

Only radar-guided sites belong in the list. Optically/IR-guided SHORAD (SA-13,
MANPADS) has nothing to shut down, and putting a mixed convoy in here would
make the whole column hold fire on every HARM shot.

Design rule (mirrors `core/air_defense.py` / `core/tasking.py`): built pydcs
groups in, one trigger out. The mission owns the wording of the radio calls;
pass a `VoiceSynth` to get them spoken as well as printed.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Optional, Sequence, Union

import structlog
from dcs import triggers

from dcs_mission_creator.core import lua

if TYPE_CHECKING:
    from dcs.mission import Mission
    from dcs.unitgroup import VehicleGroup

    from dcs_mission_creator.core.tts import VoiceSynth

log = structlog.get_logger(__name__)

_SIDE = {"blue": "coalition.side.BLUE", "red": "coalition.side.RED"}


@dataclass
class ArmSite:
    """One radar site that reacts to anti-radiation fire.

    `label` is what the radio call names ("SA-6", "EWR"). `probability` is the
    chance this crew catches a given launch, `delay_s` the recognition delay
    and `shutdown_s` how long the site stays dark, both randomised per event
    within the given range. `react_range_m` is how far from the site a launch
    still gets passed down the net.
    """

    group: "VehicleGroup"
    label: str
    probability: float = 0.85
    delay_s: tuple[float, float] = (3.0, 8.0)
    shutdown_s: tuple[float, float] = (240.0, 360.0)
    react_range_m: float = 60_000.0


# The Lua handler itself lives in `core/lua/emcon.lua`; the placeholders it
# declares (`__SITES__` / `__SIDE__` / `__SPACING__`) are filled in below.
_SCRIPT = "emcon.lua"


def _lua_str(text: Optional[str]) -> str:
    """Quote `text` as a Lua string literal (or `nil`)."""
    if text is None:
        return "nil"
    escaped = text.replace("\\", "\\\\").replace('"', '\\"')
    return f'"{escaped}"'


def arm_emcon_reaction(
    m: "Mission",
    sites: Sequence[Union["VehicleGroup", ArmSite]],
    *,
    voice: Optional["VoiceSynth"] = None,
    coalition: str = "blue",
    down_call: Optional[str] = "Magic: {label} has ceased emissions, site is dark.",
    up_call: Optional[str] = "Magic: {label} radar is radiating again.",
    announce_spacing_s: float = 7.0,
    comment: str = "ARM reaction — SAM EMCON",
) -> triggers.TriggerStart:
    """Make `sites` drop emissions when an anti-radiation missile is fired at them.

    Pass plain `VehicleGroup`s for default behaviour or `ArmSite` entries to tune
    probability / delay / shutdown per site. `down_call` and `up_call` are
    `{label}` templates announced to `coalition`; with a `VoiceSynth` the same
    words are also rendered and played from the script. Every site gets its own
    call — a single shot that darkens a whole belt queues them and plays them
    `announce_spacing_s` apart rather than dropping the later ones. Returns the
    mission-start trigger carrying the generated `DoScript`.
    """
    if coalition not in _SIDE:
        raise ValueError(f"coalition must be blue/red, got {coalition!r}")
    entries = [s if isinstance(s, ArmSite) else ArmSite(s, s.name) for s in sites]
    if not entries:
        raise ValueError("arm_emcon_reaction needs at least one site")

    rows: list[str] = []
    for site in entries:
        down_text = down_call.format(label=site.label) if down_call else None
        up_text = up_call.format(label=site.label) if up_call else None
        down_sound = voice.register(m, down_text) if voice and down_text else None
        up_sound = voice.register(m, up_text) if voice and up_text else None
        rows.append(
            "    {{name={name}, prob={prob:.3f}, delayMin={dmin:.1f}, "
            "delayMax={dmax:.1f}, downMin={smin:.1f}, downMax={smax:.1f}, "
            "range={rng:.1f}, downText={dtext}, upText={utext}, "
            "downSound={dsound}, upSound={usound}}},".format(
                name=_lua_str(site.group.name),
                prob=site.probability,
                dmin=site.delay_s[0],
                dmax=site.delay_s[1],
                smin=site.shutdown_s[0],
                smax=site.shutdown_s[1],
                rng=site.react_range_m,
                dtext=_lua_str(down_text),
                utext=_lua_str(up_text),
                dsound=_lua_str(down_sound),
                usound=_lua_str(up_sound),
            )
        )

    script = lua.render(
        _SCRIPT,
        SITES="\n".join(rows),
        SIDE=_SIDE[coalition],
        SPACING=f"{announce_spacing_s:.1f}",
    )
    rule = triggers.TriggerStart(comment=comment)
    rule.add_action(lua.InlineDoScript(script))
    m.triggerrules.triggers.append(rule)
    log.debug("armed EMCON reaction", sites=[s.group.name for s in entries])
    return rule
